codec-http2/src/main/java/io/netty/handler/codec/http2/Http2MultiplexCodec.java

```diff
@@ -66,412 +66,825 @@
-            if (eventLoop.inEventLoop()) {
-                onStreamClosed0(childChannel);
-            } else {
-                eventLoop.execute(new Runnable() {
-                    @Override
-                    public void run() {
-                        onStreamClosed0(childChannel);
-                    }
-                });
-            }
+            childChannel.streamClosed();
         }
     }
 
-    private void onStreamClosed0(Http2StreamChannel childChannel) {
-        assert childChannel.eventLoop().inEventLoop();
+    @Override
+    final void onHttp2StreamActive(ChannelHandlerContext ctx, Http2FrameStream stream) {
+        DefaultHttp2StreamChannel childChannel = new DefaultHttp2StreamChannel(stream);
 
-        childChannel.onStreamClosedFired = true;
-        childChannel.fireChildRead(CLOSE_MESSAGE);
-    }
+        childChannel.pipeline().addLast(inboundStreamHandler);
 
-    void flushFromStreamChannel() {
-        EventExecutor executor = ctx.executor();
-        if (executor.inEventLoop()) {
-            flush(ctx);
+        ChannelFuture future = ctx.channel().eventLoop().register(childChannel);
+        if (future.isDone()) {
+            registerDone(future);
         } else {
-            Runnable task = flushTask;
-            if (task == null) {
-                task = flushTask = new Runnable() {
-                    @Override
-                    public void run() {
-                        flush(ctx);
-                    }
-                };
-            }
-            executor.execute(task);
+            future.addListener(CHILD_CHANNEL_REGISTRATION_LISTENER);
         }
     }
 
-    void writeFromStreamChannel(Object msg, boolean flush) {
-        writeFromStreamChannel(msg, ctx.newPromise(), flush);
+    // TODO: This is most likely not the best way to expose this, need to think more about it.
+    final Http2StreamChannel newOutboundStream() {
+        return new DefaultHttp2StreamChannel(newStream());
     }
 
-    void writeFromStreamChannel(final Object msg, final ChannelPromise promise, final boolean flush) {
-        EventExecutor executor = ctx.executor();
-        if (executor.inEventLoop()) {
-            writeFromStreamChannel0(msg, flush, promise);
-        } else {
-            try {
-                executor.execute(new Runnable() {
-                    @Override
-                    public void run() {
-                        writeFromStreamChannel0(msg, flush, promise);
-                    }
-                });
-            } catch (Throwable cause) {
-                promise.setFailure(cause);
+    @Override
+    final void onHttp2FrameStreamException(ChannelHandlerContext ctx, Http2FrameStreamException cause) {
+        Http2FrameStream stream = cause.stream();
+        DefaultHttp2StreamChannel childChannel = ((Http2MultiplexCodecStream) stream).channel;
+
+        if (childChannel == null)  {
+            // TODO: Should we log this ?
```

Comment:
![INFO](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-info.png) Complete the task associated to this TODO comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1135)